improvements for bgp router ID ip pool

liyihuang opened the pull request:
This is the enhacenment PR of https://github.com/cilium/cilium/pull/38300 

- Enhances the router ID override logic for RouterIDIPPool mode so it can update the exsiting allocation accordingly.
- Fix my previous unit test of using incorrect return with EventuallyWithT and other errors in previous test codes in https://github.com/cilium/cilium/pull/38300
- Update the BGP docs to reflect the new RouterIDIPPool mode.



```release-note
Fix a BGP bug where the routerID specified in a CiliumBGPNodeConfigOverride was not correctly updated in RouterIDIPPool mode.
```

liyihuang committed:
docs: add bgp IP pool router-id allocation mode

This is the follow up PR of https://github.com/cilium/cilium/pull/38300
where I didn't upadte the docs after implementing the feature

Signed-off-by: Liyi Huang <liyi.huang@isovalent.com>

joestringer commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -997,13 +997,18 @@ RouterID
+and we can rely on the uniqueness of the IPv4 address to make the Router ID unique. When running in an IPv6 single-stack, 
+the lower 32 bits of MAC address of ``cilium_host`` interface are used as Router ID. 
+
+In ``ip-pool`` mode, users will need to provide an IPv4 IP pool to Cilium through the helm value 
```
```suggestion
In ``ip-pool`` mode, you must to provide an IPv4 IP pool to Cilium through the helm value 
```

liyihuang replied:
thanks. I will update it in my next push

joestringer commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -997,13 +997,18 @@ RouterID
+In ``ip-pool`` mode, users will need to provide an IPv4 IP pool to Cilium through the helm value 
+``bgpControlPlane.routerIDAllocation.ipPool``. Cilium will then assign Router IDs to BGP instances from this configured pool.
+
+If the auto assignment of the Router ID is not desired, the administrator needs to manually define it.
```
```suggestion
If the auto assignment of the Router ID is not desired, you must manually define it.
```

joestringer reviewed (changes requested):
Couple of comments from  @cilium/docs-structure perspective to ensure consistency with docs style.

YutaroHayakawa commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -997,13 +997,18 @@ RouterID
+``bgpControlPlane.routerIDAllocation.ipPool``. Cilium will then assign Router IDs to BGP instances from this configured pool.
+
+If the auto assignment of the Router ID is not desired, the administrator needs to manually define it.
 In order to configure custom Router ID, you can set ``routerID`` field in an IPv4 address format.
```
Looking at the implementation, we're not handling this case gracefully.

https://github.com/cilium/cilium/blob/038e0a74532cc9304483f0e779fa7d17ada9ee2e/operator/pkg/bgpv2/cluster.go#L385

Here we apply override, but don't check if that ID is already allocated or not. Also it doesn't record the allocated ID, so if user manually specify an ID from the pool range, ID allocator may allocate the same router ID later. I think it's better to not mention this use case or even say it's not supported explicitly until we get rid of this issue.

liyihuang replied:
Thanks for pointing it out.

The manual router ID is part of original doc https://docs.cilium.io/en/stable/network/bgp-control-plane/bgp-control-plane-v2/#bgp-configuration-override . I will push another commit to fix it and rename this PR.

yushoyamaguchi replied:
@liyihuang 
Will you write this limitation in doc as specification?
Otherwise, will you support manual ID overriding by recording allocated ID?
(This is just my interest, I'm sorry. )

liyihuang replied:
No worry. I didn't make it clear in my comments. I will fix my code and support the override. I will push another commit in this PR and rename the PR to support the override.

liyihuang replied:
I have found my orginal PR is not correct while working on my commit. that's why I have another commit to fix my origianl unit test

liyihuang committed:
fix(tests): Update BGP cluster config tests

This commit is to fix my previous unit test of using incorrect return with EventuallyWithT and other errors in previous test codes.

Signed-off-by: Liyi Huang <liyi.huang@isovalent.com>

liyihuang committed:
bgp: Enhance router ID allocation logic and update tests

Enhance the router ID override logic for RouterIDIPPool mode so it can update the exsiting allocation accordingly.
Move the restore logic to initializeJobs so the overall logic is clear

Update the unit test for override to cover new function `handleRouterIDOverride`

Signed-off-by: Liyi Huang <liyi.huang@isovalent.com>

yushoyamaguchi reviewed (approved):
Integrating these commits (777ecffc/7c0046e/406a1d2), handling of `CiliumBGPNodeConfigOverride` about _router-id_ works well in my local environment.

YutaroHayakawa commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -507,3 +530,44 @@ func ownerClusterConfigName(owners []meta_v1.OwnerReference) string {
+		return nil
+	}
+
+	// Router ID is within the pool range, check if it's already allocated to a different instance
+	if allocatedKey, exists := b.bgpRouterIDIPPool.Get(*overrideRouterID); exists {
+		if allocatedKey != currentRouterIDKey {
+			return fmt.Errorf("router ID %s is already allocated to %s, cannot use for node %s", overrideRouterID, allocatedKey, nodeName)
+		}
+		return nil
+	}
```
We allow setting arbitrary routerID in the `default` mode even if it is already used somewhere else. On the other hand, in your logic, the overridden routerID must be within the specified range and allocation fails if it is already used. I think this logic is fair, but it should be clearly stated in the document and commit message.

liyihuang replied:
Sure.  Do you think its worthy to document why we have such limit in this mode but no limit on default mode?

The reason I put such limit is because we have a pool we need to manage in ip pool mode.

 I will wait for Joe's comment together and push the changes together.

YutaroHayakawa replied:
> Sure. Do you think its worthy to document why we have such limit in this mode but no limit on default mode?

Yeah, I think the rule "allocation fails if the ID is already used" is pretty easy to reason about. The point of using IP pool mode is to strictly avoid ID duplication. The "overridden routerID must be within the specified range" is for more like an implementation simplicity in my understanding (which makes sense). Maybe we can document former.

Although, I was wondering what to do if user want to assign specific ID to specific node, but the ID was taken by auto allocation. In that case, there's no good way of reclaiming the allocated ID now.

liyihuang replied:
> The "overridden routerID must be within the specified range" is for more like an implementation simplicity in my understanding (which makes sense). Maybe we can document former.

maybe I dont understand it correctly. I dont have such limit.

If the router ID is not in the range. it will just follow the logic here https://github.com/cilium/cilium/pull/40340/files#diff-61570a61e3df66768d7af108b0ecab9a72b0f3efafd1bfc41efd53dcfbf48ec6R542-R549. we just need to free the current allocation.

https://github.com/cilium/cilium/pull/40340/files/7c0046e268c32b5854c7a4de048ee88654531c33..406a1d2a5117a43184192c3e6c84e73e4cacf7e3#diff-61570a61e3df66768d7af108b0ecab9a72b0f3efafd1bfc41efd53dcfbf48ec6R408 will just do the override


> Although, I was wondering what to do if user want to assign specific ID to specific node, but the ID was taken by auto allocation. In that case, there's no good way of reclaiming the allocated ID now.

I have to admit that's the logic that I'm not dealing with. Let's see how people will use this feature and if there are new issues reported, I can try to enhance that

liyihuang replied:
In my test here https://github.com/cilium/cilium/pull/40340/files/7c0046e268c32b5854c7a4de048ee88654531c33..406a1d2a5117a43184192c3e6c84e73e4cacf7e3#diff-fd9cf0ff9e4c36ce752fdf415037e490612f56d6de75f25fc6657b91530b607bR1474-R1488

You can see the range is 10.0.0.0/24 and I can use 192.168.1.1 as the router ID

YutaroHayakawa replied:
> maybe I dont understand it correctly. I dont have such limit.

Ah, ok sorry I was overlooking that. Then we need to fix [here](https://github.com/cilium/cilium/blob/406a1d2a5117a43184192c3e6c84e73e4cacf7e3/operator/pkg/bgpv2/cluster.go#L149-L157) as well. When we have an out-of-range routerID already allocated, we try to restore it into the pool here, but it will always treated as an error. We should ignore the out-of-range override here.

> I have to admit that's the logic that I'm not dealing with. Let's see how people will use this feature and if there are new issues reported, I can try to enhance that

Fine, we can see how it goes.

liyihuang replied:
you are absolutely right and I didn't consider the overall logic for this and will change it in my next commit.

I just would like to confirm what we should document here since you overlooked it. I guess we will document in the default mode, you can override whaterver router ID you want. In the ip pool mode, you can't have the overlap router ID within the range. Is it correct?

liyihuang replied:
resubmit the commits and add the doc. please let me know if you have any questions

joestringer commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -997,13 +997,18 @@ RouterID
+and we can rely on the uniqueness of the IPv4 address to make the Router ID unique. When running in an IPv6 single-stack, 
+the lower 32 bits of MAC address of ``cilium_host`` interface are used as Router ID. 
+
+In ``ip-pool`` mode, you must to provide an IPv4 IP pool to Cilium through the helm value 
```
```suggestion
In ``ip-pool`` mode, you must provide an IPv4 IP pool to Cilium through the helm value 
```

joestringer reviewed (commented):
Minor nit for readability. I defer to @YutaroHayakawa on the content / logical changes.

joestringer reviewed (approved):
From a docs perspective I'm OK with this. The style doesn't fully follow the best practices for docs just yet, but you're following the style that was there in the existing text. Since this is fixing a bug anyway I think that's fine. The communication comes across clear to me, which is what's most important.

YutaroHayakawa commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -149,6 +149,12 @@ func (b *BGPResourceManager) upsertNodeConfigs(ctx context.Context, config *v2.C
+					start, stop := b.bgpRouterIDIPPool.Range()
+					if start.Compare(routerID) > 0 || stop.Compare(routerID) < 0 {
+						continue
+					}
```
This logic applies to the node without override right? What happens when users first create an override for manual assignment and delete it later? In that case, we need to assign new ID for that node from the pool right?

liyihuang replied:
maybe I dont get what you mean here. the logic here is to restore the router ID from old config(stored as CR in k8s etc) to operator's memory when we restart operator. The actual assigment logic is happenning here https://github.com/liyihuang/cilium/blob/964f01831fb0eecf182d9ddfd6ce67cdac84a5fb/operator/pkg/bgpv2/cluster.go#L173-L183

YutaroHayakawa replied:
Sorry I meant here was, let's say the user first create the override with out of range ID, then the NodeConfig gets the out-of-range router ID. We restart the agent and the user deletes the override during the restart. Then the desired state here is we allocate the new in-range ID for that NodeConfig and update it. Do we handle this case in our current code? I think yes, but not super certain, so I wanted to ask that.

Here I want to check the all possible cases, we have four allocation states here. `no allocation`, `in-range no-override`, `in-range override`, `out-of-range override` and there's 9 transition cases at runtime (we don't go back to no-allocation state in IP Pool mode, maybe except the ID exhaustion case, let's talk about this case later).

- no-allocation --> in-range-no-override
- no-allocation --> in-range-override
- no-allocation --> out-of-range-override
- in-range-no-override --> in-range-override
- in-range-no-override --> out-of-range-override
- in-range-override --> in-range-no-override
- in-range-override --> out-of-range-override
- out-of-range-override --> in-range-no-override
- out-of-range-override --> in-range-override

Upon restart, we need to restore the allocation state, but in our current restoration logic, there's no handling for override, this makes the code reading a bit hard. Ideally, we should explicitly separate the restoration logic (initial list + allocation), from the regular allocation path (maybe we can do it right after the Store initialization around [here](https://github.com/liyihuang/cilium/blob/964f01831fb0eecf182d9ddfd6ce67cdac84a5fb/operator/pkg/bgpv2/manager.go#L153)) with proper override handling so that we don't need to consider the restoration case in the regular allocation path.

YutaroHayakawa replied:
> we don't go back to no-allocation state in IP Pool mode, maybe except the ID exhaustion case, let's talk about this case later

I was overlooking the ID exhaustion case in the previous review. I think we need the special handling in the `out-of-range-override --> in-range-no-override` and `out-of-range-override --> in-range-override` case. If we cannot allocate the ID there, probably, we should go to the `no-allocation` state instead of keeping the previous ID as is as users are wishing to release that ID.

liyihuang replied:
First, I think I made a mistake on restoring the router ID.

It should be something like this.

```
if start.Compare(routerID) <= 0 && stop.Compare(routerID) >= 0 {
    err := b.allocateRouterID(key, &routerID)
    if err != nil {
        errs = errors.Join(errs, fmt.Errorf("failed to restore router ID for node %s: %w", node.Name, err))
        continue
    }
}
```

I will move the restore logic to what you suggested in my next commit.

liyihuang replied:
I think my current logic is good enough to handle your `out-of-range-override --> in-range-no-override` and `out-of-range-override --> in-range-override`. let's assume the restore logic is right where we can't restore the router ID to `bgpRouterIDMap` when it's out of range.

My overall logic is the following.

when it's the BGP nodes, we will always allocate the router ID from pool here https://github.com/cilium/cilium/pull/40340/files#diff-61570a61e3df66768d7af108b0ecab9a72b0f3efafd1bfc41efd53dcfbf48ec6R173-R178. Since `bgpRouterIDMap` doesn't have the RouterID when it's out-of-range-override, it will allocate the routerID in the range.

we will have `handleRouterIDOverride` to release the routerID or assign the new one.

When we move from `out-of-range-override --> in-range-no-override` it will just use the routerID we get here https://github.com/cilium/cilium/pull/40340/files#diff-61570a61e3df66768d7af108b0ecab9a72b0f3efafd1bfc41efd53dcfbf48ec6R178

When we move from `out-of-range-override --> in-range-override`, it will get into https://github.com/cilium/cilium/pull/40340/files#diff-61570a61e3df66768d7af108b0ecab9a72b0f3efafd1bfc41efd53dcfbf48ec6R410 with `currentRouterID` from assigned pool.

then we will get into https://github.com/cilium/cilium/pull/40340/files#diff-61570a61e3df66768d7af108b0ecab9a72b0f3efafd1bfc41efd53dcfbf48ec6R567-R574 where we will release the current RouterID and use the override.


Do I miss something here?

YutaroHayakawa replied:
Sorry for the late response. I was missing this.

> Upon restart, we need to restore the allocation state, but in our current restoration logic, there's no handling for override, this makes the code reading a bit hard. Ideally, we should explicitly separate the restoration logic (initial list + allocation), from the regular allocation path (maybe we can do it right after the Store initialization around [here](https://github.com/liyihuang/cilium/blob/964f01831fb0eecf182d9ddfd6ce67cdac84a5fb/operator/pkg/bgpv2/manager.go#L153)) with proper override handling so that we don't need to consider the restoration case in the regular allocation path.

While this is still not reflected yet. I think the overall logic covers the cases I have listed. Logic-wise, now looks good to me.

YutaroHayakawa commented on operator/pkg/bgpv2/cluster_test.go:
```diff
@@ -1364,7 +1364,8 @@ func TestRouterIDAllocation(t *testing.T) {
 			InitClusterConfigs: []*v2.CiliumBGPClusterConfig{
 				{
-					ObjectMeta: meta_v1.ObjectMeta{Name: "cluster1-config"},
+					// Fake client doesn't set UID. Assign it manually to deal with multiple cluster configs
```
Sorry, I don't have a context around this UID issue. Why do we need this?

liyihuang replied:
I didn't look into to see why we need to set it. Initially, I found 30% of unit test of my multiple cluster timeout even I increate the timeout to 20s. Then I found here(https://github.com/cilium/cilium/pull/40340/files/964f01831fb0eecf182d9ddfd6ce67cdac84a5fb#diff-fd9cf0ff9e4c36ce752fdf415037e490612f56d6de75f25fc6657b91530b607bR1095-R1100). I just followed it and found the issue is gone.

YutaroHayakawa reviewed (changes requested):
Did another round of review. Regarding the update for the overriding, seems like we still need to handle some corner case.

YutaroHayakawa reviewed (approved):
I think now we're in a good shape. Thanks for your patience. We still have a code clarity issue I raised in [this comment](https://github.com/cilium/cilium/pull/40340/files#r2220035820), but the logic looks correct to me now, so I won't block this PR further.
